### PR TITLE
Getter/setter

### PR DESCRIPTION
**Pull request by milliorn**
[Added all paired Get/Set functions from this list.
](https://github.com/nwn-dotnet/NWN.Managed/issues/27)

### CODE REVIEW

**Review by Jorteck — CHANGES_REQUESTED** [review]
Looks good - only minor nit is documentation.

For class members, it should refer to itself as "this":

Use:
```cs
    /// <summary>
    /// Gets or sets the wing type of this creature.
    /// </summary>
```

Instead of:
```cs
    /// <summary>
    /// Gets or sets the wing type of the creature.
    /// </summary>
```

File: src/main/NWN/API/Object/NwCreature.cs [hunk 1]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    /// Gets a value indicating whether this creature was spawned from an encounter.
+    /// </summary>
+    public bool IsEncounter
+      => NWScript.GetIsEncounterCreature(this).ToBool();
+
+    /// <summary>
+    /// Gives xPAmount adjusted by any multiclass penalty.
+    /// </summary>
+    public void GiveXp(int xPAmount)
+      => NWScript.GiveXPToCreature(this, xPAmount);
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1182]: This does not work on NPCs, and can be moved into `NwPlayer`.

```suggestion
    /// <summary>
    /// Gives the specified XP to this player, adjusted by any multiclass penalty.
    /// </summary>
    public void GiveXp(int xPAmount)
      => NWScript.GiveXPToCreature(this, xPAmount);
```

File: src/main/NWN/API/Object/NwCreature.cs [hunk 2]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    {
+      get => (AiLevel)NWScript.GetAILevel(this);
+      set => NWScript.SetAILevel(this, (int)(AiLevel)value);
+    }
+
+    /// <summary>
+    /// Get's ability modifier for a specified ability.
+    /// </summary>
+    public Ability AbilityModifier(Ability ability)
+      => (Ability)NWScript.GetAbilityModifier((int)ability, this);
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1197]: Already implemented.

File: src/main/NWN/API/Object/NwSound.cs [hunk 3]
```diff
@@ -26,5 +26,11 @@ public sbyte Volume
     /// Stops this sound object from playing.
     /// </summary>
     public void Stop() => NWScript.SoundObjectStop(this);
+
+    /// <summary>
+    /// Gets the length of a wave (audio) file in seconds.
+    /// </summary>
+    public float DialogSoundLength(int strRef)
+      => NWScript.GetDialogSoundLength(strRef);
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwSound.cs:34]: This is a tricky function to categorize - I don't think this is the place for it though.

Are there any other functions that use .tlk indices?

File: src/main/NWN/API/Object/NwCreature.cs [hunk 4]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+      => NWScript.GiveXPToCreature(this, xPAmount);
+
+    /// <summary>
+    /// Gets or sets the AI (artificial intelligence) level.
+    /// </summary>
+    public AiLevel Ai
+    {
+      get => (AiLevel)NWScript.GetAILevel(this);
+      set => NWScript.SetAILevel(this, (int)(AiLevel)value);
+    }
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1191]: Already implemented.

File: src/main/NWN/API/Object/NwCreature.cs [hunk 5]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+      => NWScript.GetFamiliarName(creature);
+
+    /// <summary>
+    /// Gets or sets the footstep type.
+    /// </summary>
+    public FootstepType Footstep
+    {
+      get => (FootstepType)NWScript.GetFootstepType(this);
+      set => NWScript.SetFootstepType((int)value, this);
+    }
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1264]: Already implemented.

File: src/main/NWN/API/Object/NwCreature.cs [hunk 6]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+
+    /// <summary>
+    /// Gets the constant of the spell that applied the effect (SPELL_*).
+    /// </summary>
+    public Spell EffectSpellId(Effect effect)
+      => (Spell)NWScript.GetEffectSpellId(effect);
+
+    /// <summary>
+    /// Gets familiar's name.
+    /// </summary>
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1253]:
```suggestion
    /// <summary>
    /// Gets this creature's familiar name.
    /// </summary>
```

File: src/main/NWN/API/Object/NwCreature.cs [hunk 7]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    /// Gets the action mode status.
+    /// </summary>
+    public ActionMode GetActionMode(NwCreature creature, ActionMode mode)
+      => (ActionMode)NWScript.GetActionMode(creature, (int)mode);
+
+    /// <summary>
+    /// Sets the action mode status.
+    /// </summary>
+    public void SetActionMode(NwCreature creature, ActionMode mode, bool status)
+      => NWScript.SetActionMode(creature, (int)mode, status.ToInt());
```
Comment (Jorteck) [inline src/main/NWN/API/Object/NwCreature.cs:1209]: Already implemented.

**Review by Jorteck — CHANGES_REQUESTED** [review]

**Comment by Jorteck** [comment]
There's a bit of double up from this PR: https://github.com/nwn-dotnet/NWN.Managed/pull/30/files

Would you be able to merge in the latest master, and remove duplicates?

File: src/main/NWN/API/Object/NwCreature.cs [hunk 8]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    /// </summary>
+    public CreatureTailType TailType
+    {
+      get => (CreatureTailType)NWScript.GetCreatureTailType(this);
+      set => NWScript.SetCreatureTailType((int)value, this);
+    }
+
+    /// <summary>
+    /// Get a creature's animal companion creature type.
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1135]: Updated to this description.

File: src/main/NWN/API/Object/NwCreature.cs [hunk 9]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    /// Retrieves a random talent from a group of talents possessed.
+    /// </summary>
+    public TalentCategory TalentRandom(TalentCategory category)
+      => (TalentCategory)NWScript.GetCreatureTalentRandom((int)category, this);
+
+    /// <summary>
+    /// Gets the PC's familiar type.
+    /// </summary>
+    public FamiliarCreatureType FamiliarType
+      => (FamiliarCreatureType)NWScript.GetFamiliarCreatureType(this);
```
[the remark this answers is not in the document]
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1170]: Updated the summary to your suggestion.  I'll raise an issue with this function an note that it needs to be tested.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 1 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1182]: Moved to `NwPlayer`.  Summary updated.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 2 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1197]: Removed.  Thanks for spotting that.

File: src/main/NWN/API/Object/NwSound.cs (hunk 3 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwSound.cs:34]: `don't think this is the place for it though.`
I'm not sure myself.  I made the best guess I could based on the flow chart I referenced.

`Are there any other functions that use .tlk indices?`
Again I'm not sure.  I personally never used this function in nwscript or seen it implemented.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 4 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1191]: Removed.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 5 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1264]: Removed.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 6 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1253]: Summary updated.

File: src/main/NWN/API/Object/NwCreature.cs [hunk 10]
```diff
@@ -1111,5 +1111,222 @@ public CreatureModelType GetCreatureBodyPart(CreaturePart creaturePart)
...
+    /// </summary>
+    public bool ExploresMinimap
+    {
+      get => NWScript.GetCreatureExploresMinimap(this).ToBool();
+      set => NWScript.SetCreatureExploresMinimap(this, value.ToInt());
+    }
+
+    /// <summary>
+    /// Gets the default level up package.
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1150]: Summary updated.

File: src/main/NWN/API/Object/NwCreature.cs (hunk 7 above)
Comment (milliorn) [reply src/main/NWN/API/Object/NwCreature.cs:1209]: Removed.

File: src/main/NWN/API/Object/NwTrappable.cs [hunk 11]
```diff
@@ -126,5 +126,10 @@ public void SetTrapDetectedBy(bool detected, params NwCreature[] creatures)
     /// Disables this trap as if a creature disarmed it (calling the OnDisarm event respectively).
     /// </summary>
     public void DisableTrap() => NWScript.SetTrapDisabled(this);
+
+    /// <summary>
+    /// Gets who last disarmed a trapped trigger, door or placeable object.
+    /// </summary>
+    public NwTrappable LastDisarmed => (NwTrappable)NWScript.GetLastDisarmed().ToNwObject<NwGameObject>();
```
[the remark this answers is not in the document]
Comment (milliorn) [reply src/main/NWN/API/Object/NwTrappable.cs:133]: I was told this was to be used in events an instead to remove it for now.

File: src/main/NWN/API/Object/NwStationary.cs [hunk 12]
```diff
@@ -96,5 +96,19 @@ public override float Rotation
...
+    public string KeyRequiredFeedback
+    {
+      get => NWScript.GetKeyRequiredFeedback(this);
+      set => NWScript.SetKeyRequiredFeedback(this, value);
+    }
+
+    /// <summary>
+    /// Gets the last object that locked the caller.
+    /// </summary>
+    public NwStationary LastLocked => (NwStationary)NWScript.GetLastLocked().ToNwObject<NwGameObject>();
```
[the remark this answers is not in the document]
Comment (milliorn) [reply src/main/NWN/API/Object/NwStationary.cs:112]: Same as `GetLastDisarmed`.  Removed.

File: src/main/NWN/API/Object/NwTrappable.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (urothis) [reply src/main/NWN/API/Object/NwTrappable.cs]: Multiple things including doors can be trapped, wouldn't this fall under NwTrappable?

File: src/main/NWN/API/Object/NwStationary.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (urothis) [reply src/main/NWN/API/Object/NwStationary.cs]: Multiple things including doors can be locked, wouldn't this fall under NwTrappable?
As the event does not need to know who locked it, we need to get who locked it from the object type.

**Pull request merged**